### PR TITLE
Dynamodb-enhanced [preview]: Updates all operations with consumer sty…

### PR DESCRIPTION
**Pull request by cenedhryn**
…le invocations

## Description
Adds a second method to the client/table/index interfaces for all operations, that take builder consumers as input. This allows invocations in the following style:
`customerTable.getItem(r -> r.key(key));`

## Motivation and Context
Improves discoverability for the enhanced client by adding consumer-style methods for all client, table and index operations. 

## Testing
Tested with mvn clean install

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/model/WriteBatch.java [hunk 1]
```diff
@@ -82,36 +81,57 @@ public int hashCode() {
...
         public Builder<T> addPutItem(PutItemEnhancedRequest<T> request) {
             itemSupplierList.add(() -> generateWriteRequest(() -> mappedTableResource, PutItemOperation.create(request)));
             return this;
         }
 
+        public Builder<T> addPutItem(Consumer<PutItemEnhancedRequest.Builder<T>> requestConsumer) {
+            PutItemEnhancedRequest.Builder<T> builder = PutItemEnhancedRequest.builder(this.itemClass);
+            requestConsumer.accept(builder);
+            itemSupplierList.add(() -> generateWriteRequest(() -> mappedTableResource, PutItemOperation.create(builder.build())));
+            return this;
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/model/WriteBatch.java]: Could we call the other overloaded method and not duplicate code here

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/model/WriteBatch.java [hunk 2]
```diff
@@ -123,4 +143,13 @@ private WriteRequest generateWriteRequest(Supplier<MappedTableResource<T>> mappe
...
     }
+
+    private List<WriteRequest> getListIfExist(List<Supplier<WriteRequest>> itemSupplierList) {
+        if (itemSupplierList == null || itemSupplierList.isEmpty()) {
+            return null;
+        }
+        return Collections.unmodifiableList(itemSupplierList.stream()
+                                                            .map(Supplier::get)
+                                                            .collect(Collectors.toList()));
+    }
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/model/WriteBatch.java]: Can we generify and put in utils

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/DynamoDbEnhancedAsyncClient.java [hunk 3]
```diff
@@ -49,18 +50,35 @@
         throw new UnsupportedOperationException();
     }
 
+    default SdkPublisher<BatchGetResultPage> batchGetItem(Consumer<BatchGetItemEnhancedRequest.Builder> requestConsumer) {
+        throw new UnsupportedOperationException();
+    }
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/DynamoDbEnhancedAsyncClient.java]: My IDE is showing this method as not having coverage.

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/DynamoDbEnhancedAsyncClient.java [hunk 4]
```diff
@@ -49,18 +50,35 @@
...
+
     default CompletableFuture<BatchWriteResult> batchWriteItem(BatchWriteItemEnhancedRequest request) {
         throw new UnsupportedOperationException();
     }
 
+    default CompletableFuture<BatchWriteResult> batchWriteItem(Consumer<BatchWriteItemEnhancedRequest.Builder> requestConsumer) {
+        throw new UnsupportedOperationException();
+    }
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/DynamoDbEnhancedAsyncClient.java]: My IDE is showing this method as not having coverage.

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/DynamoDbEnhancedClient.java [hunk 5]
```diff
@@ -48,18 +49,34 @@
...
+        throw new UnsupportedOperationException();
+    }
+
     default BatchWriteResult batchWriteItem(BatchWriteItemEnhancedRequest request) {
         throw new UnsupportedOperationException();
     }
 
+    default BatchWriteResult batchWriteItem(Consumer<BatchWriteItemEnhancedRequest.Builder> requestConsumer) {
+        throw new UnsupportedOperationException();
+    }
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/DynamoDbEnhancedClient.java]: My IDE is showing this method as not having coverage.

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/core/DefaultDynamoDbIndexTest.java [hunk 6]
```diff
@@ -37,6 +38,9 @@
     @Mock
     private MapperExtension mockMapperExtension;
 
+    @Mock
+    private PaginatedIndexOperation<?,?,?,?> mockPaginatedOperation;
+
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/core/DefaultDynamoDbIndexTest.java]: This does not appear to be used. Can we remove it?

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncBasicCrudTest.java [hunk 7]
```diff
@@ -228,6 +226,25 @@ public void deleteTable() {
...
+    @Test
+    public void putItemInterfaceToOperationSucceeds() {
+        Record record = new Record()
+            .setId("id-value")
+            .setSort("sort-value")
+            .setAttribute("one");
+
+        mappedTable.putItem(Record.class, r -> r.item(record)).join();
+        mappedTable.putItem(PutItemEnhancedRequest.builder(Record.class).item(record).build()).join();
+    }
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncBasicCrudTest.java]: I think we should be able to make basic assertions here beyond 'it didn't blow up'.

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncBasicCrudTest.java [hunk 8]
```diff
@@ -376,11 +381,17 @@ public void putWithConditionThatFails() {
...
     }
 
+    @Test
+    public void deleteItemInterfaceToOperationSucceeds() {
+        mappedTable.deleteItem(r -> r.key(Key.create(stringValue("id-value"), stringValue("sort-value")))).join();
+        mappedTable.deleteItem(DeleteItemEnhancedRequest.builder()
+                                                        .key(Key.create(stringValue("id-value"), stringValue("sort-value")))
+                                                        .build()).join();
+    }
+
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncBasicCrudTest.java]: I think we should be able to make basic assertions here beyond 'it didn't blow up'.

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncBasicCrudTest.java [hunk 9]
```diff
@@ -436,6 +451,17 @@ public void deleteWithConditionThatFails() {
...
+    public void updateItemInterfaceToOperationSucceeds() {
+        Record record = new Record()
+            .setId("id-value")
+            .setSort("sort-value")
+            .setAttribute("one");
+
+        mappedTable.updateItem(Record.class, r -> r.item(record)).join();
+        mappedTable.updateItem(UpdateItemEnhancedRequest.builder(Record.class).item(record).build()).join();
+    }
+
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncBasicCrudTest.java]: I think we should be able to make basic assertions here beyond 'it didn't blow up'.

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncIndexScanTest.java [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import software.amazon.awssdk.core.async.SdkPublisher;
+import software.amazon.awssdk.extensions.dynamodb.mappingclient.DynamoDbAsyncIndex;
+import software.amazon.awssdk.extensions.dynamodb.mappingclient.DynamoDbAsyncTable;
+import software.amazon.awssdk.extensions.dynamodb.mappingclient.DynamoDbEnhancedAsyncClient;
+import software.amazon.awssdk.extensions.dynamodb.mappingclient.DynamoDbEnhancedClient;
+import software.amazon.awssdk.extensions.dynamodb.mappingclient.DynamoDbIndex;
+import software.amazon.awssdk.extensions.dynamodb.mappingclient.DynamoDbTable;
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncIndexScanTest.java]: Unused imports

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncIndexScanTest.java [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+        BufferingSubscriber<T> subscriber = new BufferingSubscriber<>();
+        publisher.subscribe(subscriber);
+        subscriber.waitForCompletion(1000L);
+
+        assertThat(subscriber.isCompleted(), is(true));
+        assertThat(subscriber.bufferedError(), is(nullValue()));
+        assertThat(subscriber.bufferedItems().size(), is(expectedNumberOfResults));
+
+        return subscriber.bufferedItems();
+    }
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncIndexScanTest.java]: I think we're using this code in a bunch of places, can we refactor it into a base test class or util?

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncIndexScanTest.java [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+    @Test
+    public void scanAllRecordsDefaultSettings() {
+        insertRecords();
+
+        SdkPublisher<Page<Record>> publisher = keysOnlyMappedIndex.scan(ScanEnhancedRequest.builder().build());
+        List<Page<Record>> results = drainPublisher(publisher, 1);
+        Page<Record> page = results.get(0);
+
+        assertThat(page.items(), is(KEYS_ONLY_RECORDS));
+    //    assertThat(page.lastEvaluatedKey(), is(nullValue()));
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncIndexScanTest.java]: Why is this commented out?

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/BasicCrudTest.java [hunk 13]
```diff
@@ -222,6 +220,25 @@ public void deleteTable() {
...
+    public void putItemInterfaceToOperationSucceeds() {
+        Record record = new Record()
+            .setId("id-value")
+            .setSort("sort-value")
+            .setAttribute("one");
+
+        mappedTable.putItem(Record.class, r -> r.item(record));
+        mappedTable.putItem(PutItemEnhancedRequest.builder(Record.class).item(record).build());
+    }
+
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/BasicCrudTest.java]: I think we should be able to make basic assertions here beyond 'it didn't blow up'.

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/BasicCrudTest.java [hunk 14]
```diff
@@ -360,10 +371,17 @@ public void putWithConditionThatFails() {
...
     }
 
+    @Test
+    public void deleteItemInterfaceToOperationSucceeds() {
+        mappedTable.deleteItem(r -> r.key(Key.create(stringValue("id-value"), stringValue("sort-value"))));
+        mappedTable.deleteItem(DeleteItemEnhancedRequest.builder()
+                                                        .key(Key.create(stringValue("id-value"), stringValue("sort-value")))
+                                                        .build());
+    }
+
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/BasicCrudTest.java]: I think we should be able to make basic assertions here beyond 'it didn't blow up'.

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/BasicCrudTest.java [hunk 15]
```diff
@@ -418,6 +436,17 @@ public void deleteWithConditionThatFails() {
...
+    public void updateItemInterfaceToOperationSucceeds() {
+        Record record = new Record()
+            .setId("id-value")
+            .setSort("sort-value")
+            .setAttribute("one");
+
+        mappedTable.updateItem(Record.class, r -> r.item(record));
+        mappedTable.updateItem(UpdateItemEnhancedRequest.builder(Record.class).item(record).build());
+    }
+
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/BasicCrudTest.java]: I think we should be able to make basic assertions here beyond 'it didn't blow up'.

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/BasicQueryTest.java [hunk 16]
```diff
@@ -25,6 +25,8 @@
 import static software.amazon.awssdk.extensions.dynamodb.mappingclient.model.QueryConditional.equalTo;
 import static software.amazon.awssdk.extensions.dynamodb.mappingclient.staticmapper.AttributeTags.primaryPartitionKey;
 import static software.amazon.awssdk.extensions.dynamodb.mappingclient.staticmapper.AttributeTags.primarySortKey;
+import static software.amazon.awssdk.extensions.dynamodb.mappingclient.staticmapper.AttributeTags.secondaryPartitionKey;
+import static software.amazon.awssdk.extensions.dynamodb.mappingclient.staticmapper.AttributeTags.secondarySortKey;
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/BasicQueryTest.java]: Unused imports.

File: services-custom/dynamodb-enhanced/README.md [hunk 17]
```diff
@@ -80,66 +80,109 @@ values used are also completely arbitrary.
...
+named after. The examples below are the most simple variants of each
+operation possible, using the the two styles available for constructing
+requests with either builder or consumers. These commands can be 
+customized by using the builders provided for each command and offer 
+most of the features available in the low-level DynamoDB SDK client.
 
    ```java
    // CreateTable
-   customerTable.createTable(CreateTableEnhancedRequest.create());
+   customerTable.createTable(() -> {});
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/README.md]: I'm not sure I like this, speaking to Matt he said there is prior art for having an overloaded version of this that takes no arguments at all. So-called 'simple' method form.

File: services-custom/dynamodb-enhanced/README.md [hunk 18]
```diff
@@ -80,66 +80,109 @@ values used are also completely arbitrary.
...
    // Query
-   Iterable<Page<Customer>> customers = customerTable.query(QueryEnhancedRequest.create(equalTo(Key.create(stringValue("a123")))));
-   
+   Iterable<Page<Customer>> customers = customerTable.query(r -> r.queryConditional(equalTo(Key.create(stringValue("a123")))));
+   Iterable<Page<Customer>> customers = customerTable.query(QueryEnhancedRequest.builder()
+                                                                                .queryConditional(equalTo(Key.create(stringValue("a123"))))
+                                                                                .build());
    // Scan
-   Iterable<Page<Customer>> customers = customerTable.scan(ScanEnhancedRequest.create());
+   Iterable<Page<Customer>> customers = customerTable.scan(() -> {});
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/README.md]: Can we do the 'simple method' implementation here with no arguments.

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncBasicCrudTest.java (hunk 7 above)
Comment (cenedhryn) [reply services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncBasicCrudTest.java]: I could not put my heart into the wiring tests after all and I removed them. All wiring combinations are still tested (no unused interface methods). We can always rethink later.

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/core/Utils.java [hunk 19]
```diff
@@ -117,4 +119,13 @@ public static String cleanAttributeName(String key) {
         return sortKeyValue.map(attributeValue -> Key.create(partitionKeyValue, attributeValue))
                            .orElseGet(() -> Key.create(partitionKeyValue));
     }
+
+    public static <T> List<T> getListIfExist(List<Supplier<T>> itemSupplierList) {
```
Comment (bmaizels) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/core/Utils.java]: Now it's generic and global we should rename this to be a little more specific to what it does, which is call each supplier and make a list of the results.

Comment (cenedhryn) [reply services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/core/Utils.java]: yes, definitely. Updated.

**Comment by ocind** [comment]
hi @cenedhryn I was following the tracking issue for DynamoDb Mapper in V2 [here](https://github.com/aws/aws-sdk-java-v2/issues/35). I assume that DynamoDb-enhanced would be released as part of the final DynamoDbMapper? If that is the case, I think it would be good if we can reference the tracking [issue](https://github.com/aws/aws-sdk-java-v2/issues/35) inside PRs related to DynamoDb-enhanced. I saw that there are lots of PRs related to this recently and I think it's great to give a sense of progress on the tracking issue. 

I am sure there are a lot of people (like myself) which are following the tracking issue since I would say high-level dynamodb client is a pretty significant thing that is missing from V2 sdk currently.

**Comment by bmaizels** [comment]
@ocind 
> I am sure there are a lot of people (like myself) which are following the tracking issue since I would say high-level dynamodb client is a pretty significant thing that is missing from V2 sdk currently.

You got it. We're making a big push to launch this. I'm making a point of tagging all the PRs against that issue. You'll be seeing a lot of activity there over the next few weeks. Thanks for calling it out.

**Pull request merged**